### PR TITLE
Feature/CVSB-19224

### PR DESCRIPTION
**Pull request by PaulBenn-BJSS**
## CVSB-19224
Convert DynamoDB documents (NoSQL) into sets of RDS rows (SQL).

### Supported conversions:

- Technical Record Documents (see `tech-record-document-conversion.ts`):
    - Source (DynamoDB table): `Technical_Records`
    - Destination (RDS tables), in upsertion order:
        - `vehicle`
        - `make_model`
        - `vehicle_class`
        - `vehicle_subclass`
        - `identity`
        - `contact_details`
        - `technical_record`
        - `psv_brakes`
        - `axle_spacing`
        - `microfilm`
        - `plate`
        - `axle`
- Test Result Records (see `test-result-record-conversion.ts`):
    - Source (DynamoDB table): `Test_Results`
    - Destination (RDS tables), in upsertion order:
        - `vehicle`
        - `test_station`
        - `tester`
        - `vehicle_class`
        - `preparer`
        - `identity`
        - `fuel_emission`
        - `test_type`
        - `test_result`
        - `defect`
        - `location`
        - `test_defect`
        - `custom_defect`

### Procedure:

1. Receive DynamoDB stream event
2. Identify source table from event's `eventSourceARN`
3. Parse raw DynamoDB image into `DynamoDbImage`
4. Select converter based on source table
5. Parse `DynamoDbImage` to correct TS model
6. Upsert related vehicle (in separate transaction)
7. Begin transaction
8. Upsert to all relevant RDS tables
    - either a **partial** upsert (immutable data)
    - or a **full** upsert (update all fields if key matches)
9. Commit (or rollback on error)

### CODE REVIEW

**Review by SanjeetBJSS — DISMISSED** [review]

File: tests/resources/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Match Aurora MySQL version as closely as possible
+FROM mysql:5.7.33
```
Comment (ChrisPeska-BJSS) [inline tests/resources/Dockerfile:2]: Please downgrade MySql version to 5.7.12 so it does not exceed minor version number of AWS Aurora compatibility. 

>  All 2.x Aurora MySQL engine versions are wire-compatible with Community MySQL 5.7.12. 
https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/AuroraMySQL.Updates.Versions.html

Comment (PaulBenn-BJSS) [reply tests/resources/Dockerfile]: Addressed https://github.com/dvsa/cvs-tsk-update-store/pull/1/commits/6a6ed90b3eb3a0025891607808dde506ac7aec0d

File: .gitmodules [hunk 2]
```diff
@@ -0,0 +1,4 @@
+[submodule "cvs-nop"]
```
Comment (sdh100shaun) [inline .gitmodules:1]: Ah so we went down the submodule route ... 
I think it will be good to have a DVSA stance on this @SanjeetBJSS  
and bring some best practice advice in for other team members

**Review by JonLazariniDD — COMMENTED** [review]
Can we add some context around all the services fit together in the README please?
Maybe some architrecture section? Is that the correct diagram for us please?
https://wiki.dvsacloud.uk/display/HVT/VOTT+Architecture

File: .npmrc [hunk 3]
```diff
@@ -1,2 +0,0 @@
-audit-level = high
```
Comment (JonLazariniDD) [inline .npmrc:1]: Why are we removing the .npmrc config file? I think it is best to use that for our project config.
The pipeline will use that to run its scripts too. Maybe we can reuse the one we have in the update store service?

https://github.com/dvsa/cvs-svc-enquiry/blob/feature/CVSB-19222/.npmrc

File: README.md [hunk 4]
```diff
@@ -1 +1,130 @@
- # cvs-tsk-update-store
+# cvs-tsk-update-store
+
+## Overview
+Lambda handler and business logic for the CVS task "Update Store":
+
+1. Receives DynamoDB stream events, containing NoSQL document snapshots ("**images**")
+2. Converts these images to a TypeScript model, then to an SQL-friendly tuple.
+3. Upserts (inserts or updates) tuples into RDS [Aurora][aurora], a relational SQL database.
+
```
Comment (JonLazariniDD) [inline README.md:9]: Will probably need your help to run the `test-local-i` script locally please.
I can't also get the `start` script to work.
I think some dependencies are missing for me and fail to make this script run when you have time.

File: package.json [hunk 5]
```diff
@@ -7,6 +7,10 @@
     "node": "10.18.1",
     "npm": "5.6.0"
   },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/dvsa/cvs-tsk-update-store.git"
+  },
   "scripts": {
     "start": "npm run kill:offline && npm run start:offline && serverless invoke local -f main --path tests/resources/queue-event.json",
```
Comment (JonLazariniDD) [inline package.json:15]: We seem to be using the old template? Why are we not using the one Rob created?
It is not using webpack to bundle the code which has tree shaking, etc..

I think it is missing a few things DVSA require for us to start such the dvsa/eslint plugin

File: package.json [hunk 6]
```diff
@@ -18,14 +22,15 @@
     "build:docker": "docker-compose -f docker-compose.yml up -d --build",
     "test": "BRANCH=local AWS_XRAY_CONTEXT_MISSING=LOG_ERROR jest --testMatch=\"**/*.unitTest.ts\" --coverage",
     "test-i": "BRANCH=local jest --testMatch=\"**/*.intTest.ts\" --runInBand",
+    "test-local-i": "USE_CONTAINERIZED_DATABASE=1 BRANCH=local jest --testMatch=\"**/*.intTest.ts\" --runInBand",
     "tslint": "tslint src/**/*.ts tests/**/*.ts -q",
     "prepush": "npm test && npm run tslint",
     "sonar-scanner": "AWS_XRAY_CONTEXT_MISSING=LOG_ERROR BRANCH=local npm run tslint && AWS_XRAY_CONTEXT_MISSING=LOG_ERROR npm run coverage && sonar-scanner",
     "coverage": "BRANCH=local SLS_DEBUG=* jest --testMatch=\"**/*.*Test.ts\" --coverage",
     "audit-dep": "npm audit --json | node_modules/.bin/audit-filter --nsp-config=.nsprc --audit=-",
```
Comment (JonLazariniDD) [inline package.json:30]: if we add the .npmrc file we can remove this script and replace it with that `"audit": "npm audit"` which is ran in the pipeline

File: package.json [hunk 7]
```diff
@@ -18,14 +22,15 @@
     "build:docker": "docker-compose -f docker-compose.yml up -d --build",
     "test": "BRANCH=local AWS_XRAY_CONTEXT_MISSING=LOG_ERROR jest --testMatch=\"**/*.unitTest.ts\" --coverage",
     "test-i": "BRANCH=local jest --testMatch=\"**/*.intTest.ts\" --runInBand",
+    "test-local-i": "USE_CONTAINERIZED_DATABASE=1 BRANCH=local jest --testMatch=\"**/*.intTest.ts\" --runInBand",
     "tslint": "tslint src/**/*.ts tests/**/*.ts -q",
     "prepush": "npm test && npm run tslint",
```
Comment (JonLazariniDD) [inline package.json:27]: I think we can change the hooks order like we have on update store however a few things are missing i think:
- commitlint
- prettier
- eslint
- scripts around the automation of the release(s)

https://github.com/dvsa/cvs-svc-enquiry/blob/feature/CVSB-19222/package.json#L11

File: package.json [hunk 8]
```diff
@@ -53,31 +58,34 @@
...
     "husky": "^3.0.9",
-    "jest": "^24.9.0",
+    "jest": "^26.0.0",
     "jest-plugin-context": "^2.9.0",
     "jest-sonar-reporter": "^2.0.0",
     "lambda-tester": "^3.5.0",
     "serverless": "^2.24.0",
     "serverless-dependency-invoke": "^0.0.9",
-    "serverless-offline": "^5.12.0",
+    "serverless-offline": "^6.9.0",
```
Comment (JonLazariniDD) [inline package.json:88]: I think it is best if we use the save-exact when saving packages, we are not sure whether all the packages we use follow the semver convention and not locking a specific version could break our app if breaking changes were pushed on patch, minor version for example.

We can also add dependabot for upgrading packages

File: scripts/tools-setup.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
+const liquibase = require("liquibase");
+const path = require("path");
+
+// see README > "Integration tests" for why this awful code exists
+(async () => await liquibase({
+    liquibase: path.join(__dirname, "../node_modules/liquibase/lib/liquibase-4.0.0/liquibase"),
+    changeLogFile: path.join(__dirname, "../cvs-nop/changelog-master.xml"),
+    url: 'jdbc:mysql://localhost:3306/CVSBNOP',
+    username: "root",
```
Comment (JonLazariniDD) [inline scripts/tools-setup.js:9]: Couldn't we add that to environmental variables? I would suspect this would need to be changed in the pipeline based on our environments?
I think for now we seem to be having `local`, `dev`, `integration`, `preprod`, `prod`

We could use an `.env` file, I think the pipeline might have that already (@githanium )?

File: scripts/tools-setup.js [hunk 10]
```diff
@@ -0,0 +1,18 @@
+const liquibase = require("liquibase");
+const path = require("path");
+
+// see README > "Integration tests" for why this awful code exists
+(async () => await liquibase({
+    liquibase: path.join(__dirname, "../node_modules/liquibase/lib/liquibase-4.0.0/liquibase"),
```
Comment (JonLazariniDD) [inline scripts/tools-setup.js:6]: we may want to dynamically get the path as I am thinking let's say we were to change the liquidbase version we use, this script would fail

**Comment by PaulBenn-BJSS** [comment]
@JonLazariniDD Addressed [your comment](https://github.com/dvsa/cvs-tsk-update-store/pull/1#pullrequestreview-638686197) with [this commit](https://github.com/dvsa/cvs-tsk-update-store/pull/1/commits/9f17347042e1dff1b8d78015e69a1e0fe3836f8d). Let me know if you'd like a more written-up version :)

**Review by JonLazariniDD — COMMENTED** [review]
I think we need to clarify the ACs on the ticket for the testers.
So ATM, if my understanding is correct we do not have delete but have done some work for the following queries
- update
- insert

for the following services as per the design/diagram?
- test-results
- tech-records
- activities

File: package.json (hunk 5 above)
Comment (PaulBenn-BJSS) [reply package.json:15]: @SanjeetBJSS is aware of this. Basically, it was a choice between:
* Using the old template and getting the work done
* Using the new template and not getting the work done

So he picked the first approach :D the new template at the start of this project was a constantly evolving piece of software, and Rob Waring also agrees this was the right choice

Note Tanio has also edited the package.json on my branch, some changes are his.

File: package.json (hunk 6 above)
Comment (PaulBenn-BJSS) [reply package.json:30]: @githanium As far as I'm aware the current structure works for this project - have I misunderstood?

File: scripts/tools-setup.js (hunk 9 above)
Comment (PaulBenn-BJSS) [reply scripts/tools-setup.js:9]: This was added under explicit instructions from @githanium, but can be removed if he says it's OK :)

File: src/functions/process-stream-event.ts [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+            // parse source ARN
+            const eventSourceArn: EventSourceArn = stringToArn(dynamoRecord.eventSourceARN!);
+
+            console.info(`source ARN region:     '${eventSourceArn.region}'`);
+            console.info(`source ARN account ID: '${eventSourceArn.accountId}'`);
+            console.info(`source ARN timestamp:  '${eventSourceArn.timestamp}'`);
+
+            // is this an INSERT, UPDATE, or DELETE?
+            const operationType: SqlOperation = deriveSqlOperation(dynamoRecord.eventName!);
```
[the remark this answers is not in the document]
Comment (PaulBenn-BJSS) [reply src/functions/process-stream-event.ts:39]: Do you have an up-to-date version of the code? I think `queue-event.json` was deleted a while ago.

* You are correct - the SQS event does not contain any of these fields :)
* However, `event.Records[i].body` for any `i` is of type `DynamoDBRecord` (defined by aws-sdk), so it does contain them. This body is parsed on line 25 of the current handler.

File: package.json (hunk 8 above)
Comment (PaulBenn-BJSS) [reply package.json]: I'm not adding dependabot as we need to make the release cut, but feel free to change this on a future branch. However, I have set all versions to exact on [this commit](https://github.com/dvsa/cvs-tsk-update-store/pull/1/commits/62cf5de601f82c513403092b83197f755092bd5e)

File: .npmrc (hunk 3 above)
Comment (PaulBenn-BJSS) [reply .npmrc:1]: Addressed https://github.com/dvsa/cvs-tsk-update-store/pull/1/commits/c9ba1e53216fa6f1d412e93a875ecbb42caef4c2

File: scripts/tools-setup.js (hunk 10 above)
Comment (PaulBenn-BJSS) [reply scripts/tools-setup.js]: Resolved: file deleted with Tanio's blessing https://github.com/dvsa/cvs-tsk-update-store/pull/1/commits/e8fe927eb34091e84369f214815a38672a2b7f48

File: scripts/tools-setup.js (hunk 9 above)
Comment (PaulBenn-BJSS) [reply scripts/tools-setup.js]: Resolved: file deleted with Tanio's blessing https://github.com/dvsa/cvs-tsk-update-store/pull/1/commits/e8fe927eb34091e84369f214815a38672a2b7f48

**Review by SanjeetBJSS — APPROVED** [review]

**Review by SanjeetBJSS — APPROVED** [review]

File: src/models/shared-enums.ts [hunk 12]
```diff
@@ -0,0 +1,36 @@
+export type VehicleType = "psv" | "hgv" | "trl" | "car" | "lgv" | "motorcycle";
+
+export type VehicleConfiguration =
```
Comment (JonLazariniDD) [inline src/models/shared-enums.ts:3]: I need @m-ujaffer second pair of eyes on the definition of models please.
They also could be part of separate shared repo

File: src/models/tech-record-document.ts [hunk 13]
```diff
@@ -0,0 +1,28 @@
+import {DynamoDbImage, parseStringArray} from "../services/dynamodb-images";
+import {parseTechRecords, TechRecords} from "./tech-record";
+
+// https://wiki.dvsacloud.uk/pages/viewpage.action?spaceKey=HVT&title=Technical+API+Changelog
```
Comment (JonLazariniDD) [inline src/models/tech-record-document.ts:4]: We could be looking at automatically generating the interfaces from the specs in this shared module

File: src/services/connection-pool.ts [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+        const connectionPool = await getConnectionPool();
+        const [rows, fields] = await connectionPool.execute(sql, templateVariables);
+        return {rows, fields};
+    }
+};
+
+// npm packages mysql and mysql2 will throw an error on encountering JS "undefined".
+// In contrast, an explicit JS "null" translates to SQL "NULL".
+// hence, transform all "undefined" to "null".
+const undefinedToNull = (array: any[]): any[] => {
```
Comment (JonLazariniDD) [inline src/services/connection-pool.ts:61]: It is probably one for all of us and is happening because of the joy of developing with Js and perhaps unnecessary since the data seems to be flowing one way - dynamo data -> sql data which is then used for query purposes by analysts?
By transforming undefined to null we are losing its initial intent so i was thinking sql would allow undefined to be referenced as a string ? I will also need some clarity around what type of queries will be performed and use cases for people consuming the service to understand better?

File: tests/integration/database-teardown.ts [hunk 15]
```diff
@@ -0,0 +1,12 @@
+import {executeSql} from "../../src/services/connection-pool";
+import {allTables, TableDetails} from "../../src/services/table-details";
+
+export const databaseTearDown = async () => {
+    // only run if database is not containerized - no point in tearing down data on an ephemeral environment
```
Comment (JonLazariniDD) [inline tests/integration/database-teardown.ts:5]: I assume the e2e do not use that otherwise I would think when it comes to testing we might want to start from a clean slate and not preserve state across tests?

File: tests/integration/tech-record-document-conversion.intTest.ts [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+import {processStreamEvent} from "../../src/functions/process-stream-event";
+import {getConnectionPoolOptions} from "../../src/services/connection-pool-options";
+
+useLocalDb();
+
+describe("convertTechRecordDocument() integration tests", () => {
+    let container: StartedTestContainer;
+
+    beforeAll(async () => {
+        jest.setTimeout(60_000);
```
Comment (JonLazariniDD) [inline tests/integration/tech-record-document-conversion.intTest.ts:16]: This is an interesting one and not sure if optimisation could be required in the future. It is not something we can really control but I guess it didn't time out in the pipeline and built the service already?
I think some good points/questions have been raised earlier around stress/perf testing. Do you know what a prod record would look like?
We might ask Mike or Sanj but hopefully we won't expect tech-records or test-results to be to big

Comment (PaulBenn-BJSS) [reply tests/integration/tech-record-document-conversion.intTest.ts:16]: Hi, this is not the time it takes update-store to process one document. That time is about 7 seconds on average from the time the stream event is emitted from the source database to the time it reaches the destination database. This `60_000` ms is the time it takes for the docker daemon to spin up a MySQL image and for the spawned Liquibase process to apply to it. the test itself takes less than 300ms on average :)

File: tests/integration/database-teardown.ts (hunk 15 above)
Comment (PaulBenn-BJSS) [reply tests/integration/database-teardown.ts]: I have removed this file after confirming with Tanio that he no longer needs it, as he has fixed this in the pipeline. This was indeed the intention with the Dockerized database, however Tanio was hesitant to work on Docker-in-Docker so we added in both this and the `USE_CONTAINERIZED_DATABASE` variable, temporarily, to help him out.

File: tests/integration/tech-record-document-conversion.intTest.ts (hunk 16 above)
Comment (PaulBenn-BJSS) [reply tests/integration/tech-record-document-conversion.intTest.ts:16]: With respect to records similar to production - see some examples based on Deepika's regression suite under my own `src/tests/resources`:
* HGV, PSV and TRL tech records
* HGV, PSV and TRL test results

**Comment by PaulBenn-BJSS** [comment]
With respect to ignoring MySQL connector (https://github.com/dvsa/cvs-tsk-update-store/pull/1#pullrequestreview-639141585)

I don't think so, no. Generally, when you clone a repo, you want to have as much tooling as possible already inside the repo such that you can start work with less set-up / depending on less commands.

`mysql-connector` is a regular dependency in most other languages - e.g. in Java, you just add a line in Gradle or a few lines in Maven and it just works. In fact, in the case of Gradle, the entire Gradle installation also lives in the repo, such that when you clone it, it can immediately run code. Unfortunately, for TS, this model is not easy to implement:
* the connector does not exist as an **official** npm dependency
* Using a script would involve curl'ing an official mirror and sticking the dependency in your local copy of the project - this would introduce problems such as downloading a "latest" version which doesn't match the one previously used, or the download failing, or write access failing
* In addition, the dependency is difficult to limit to the `intTest` package, in such a way that it is not packaged into the production bundle. A better JS engineer could do this, maybe

In summary, I don't think having the external binary is much of a big deal here, since the tests depend on it. Let me know if you disagree

File: README.md (hunk 4 above)
Comment (PaulBenn-BJSS) [reply README.md:9]: Resolved as per our chat yesterday

**Comment by PaulBenn-BJSS** [comment]
With respect to how the work relates to existing tables (https://github.com/dvsa/cvs-tsk-update-store/pull/1#pullrequestreview-638726467))

Almost!

The following operations are supported:
* `INSERT`
* `UPDATE`

The following tables are supported:
- `technical-records`
- `test-results`

The following tables are *not* supported (for this PR):
- `activities`

File: src/models/axles.ts [hunk 17]
```diff
@@ -0,0 +1,121 @@
+import {DynamoDbImage} from "../services/dynamodb-images";
+import {Maybe} from "./optionals";
```
[the remark this answers is not in the document]
Comment (PaulBenn-BJSS) [reply src/models/axles.ts:2]: For this project, we have 3 big-bang (everything happens at the same time) conversions:
- from raw JSON to image
- from image to TS model
- from model to SQL statement + template variables

I would be quite difficult to change conversions this large to use a Ramda functional pipeline, as there is lots branching logic involved in all three steps (image selection, operation selection, manual mapping, undefined checks...). If we had more time, I think perhaps I would have chosen something like this, but IMO the code is readable enough as it is - the introduction of a 1-liner `Maybe<T>` type was way easier than a new npm :)

File: src/services/connection-pool.ts (hunk 14 above)
Comment (PaulBenn-BJSS) [reply src/services/connection-pool.ts:61]: Defining `undefined` values as the raw string `undefined` is a decision for the business intelligence department / the analysts who will actually be using this SQL. If it was me, personally, I'd want both explicit `null` and missing values to translate to SQL `NULL`. @SanjeetBJSS can I get some thoughts on this?

There is also the question of existing CVS API specs, which have two problems related to nulls:
* Some fields are required, but also nullable. Passing anything but an explicit null on these fields causes a `400 Bad Request`.
* Some fields are optional, and nullable, but passing an explicit null to them causes a `400 Bad Request`.

This is unreasonable and extremely hard to follow / implement in code (see my [`ExplicitNullAdapterFactory`](https://github.com/dvsa/cvs-auto-svc-vott/blob/feature/CVSB-19155/src/main/java/vott/models/adapter/ExplicitNullAdapterFactory.java) helping out the regression guys). I wanted to follow a simpler approach for the SQL side of things

Comment (SanjeetBJSS) [reply src/services/connection-pool.ts:61]: Hi Guys, I would like undefined set to NULL. The fact it is a string, is easily established by the metadata data, remember this is RDS now.

File: package.json (hunk 6 above)
Comment (githanium) [reply package.json:30]: @JonLazariniDD's approach allows developers to customise the behaviour of the audit process in  the repo itself. This approach will give you the ability to set the audit level as PO wish in the repo. The pipeline instead just run npm audit without much config to be modified @PaulBenn-BJSS

File: src/functions/process-stream-event.ts [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+import {destroyConnectionPool} from "../services/connection-pool";
+
+/**
+ * λ function: convert a DynamoDB document to Aurora RDS rows
+ * @param event - DynamoDB stream event, containing DynamoDB document image
+ * @param context - λ context
+ */
+export const processStreamEvent: Handler = async (event: SQSEvent, context: Context): Promise<any> => {
+    try {
+        console.info("Received SQS event: ", event);
```
[the remark this answers is not in the document]
Comment (githanium) [reply src/functions/process-stream-event.ts:15]: Please print out only the logs that are absolutely necessary to troubleshoot the service. Cloudwatch can be quite expensive when run millions of times.

[the remark this answers is not in the document]
Comment (PaulBenn-BJSS) [reply src/functions/process-stream-event.ts]: Addressed https://github.com/dvsa/cvs-tsk-update-store/pull/1/commits/d8656cc7069c7e849950cbc34ace4d42ca6b3dc0

File: package.json (hunk 5 above)
Comment (SanjeetBJSS) [reply package.json:15]: Hi both, this was a 6 week activity and the template had not been signed off (and isn't). Given we are taking a release cut today and testing is complete, I do not want to change this now. I accept the technical debt incurred and I will pop a ticket in the backlog to address it. Hopefully  we can do this when we plug the activities data in. @sdh100shaun  @bristol-mike-jones , please shout if you have a different opinion.

File: .gitmodules (hunk 2 above)
Comment (SanjeetBJSS) [reply .gitmodules:1]: Given that this services only job is to take a Dynamo Stream and convert it to the RDS's table structure ensuring that the structure is accurate and up to date is imperative . If you would like, I can get Paul to copy and paste the files in. I used to use the copy and paste approach in c#, when plugging WCF DTO's into WPF but given the nature of this service, and the simplicity of the SQL script, it kinda made sense. There is no submodules or dependancies below it, and no chance of getting into a web of dependancy. If you would like Paul to copy it in, I can ask him to.

File: package.json (hunk 7 above)
Comment (PaulBenn-BJSS) [reply package.json:27]: - `commitlint` will not allow a single one of the commit messages I have made on this repo, as well as involving an extra set-up step to make sure environment variable `HUSKY_GIT_PARAMS` is set. Perhaps consider adding this at a later date? I want to make this release cut happen :)
- `prettier`: yes, I can add this. exactly 60 files will change as a result of this commit. Let me know if you want this to happen.
- `eslint` is superseded by `tslint`, as far as I know, so i'm not adding `eslint` on top
- scripts: isn't this done already? the pipeline works fine with the existing tools-setup
- envc configuration: same story as the above - please let me know if this is not the case

File: .npmrc [hunk 19]
```diff
@@ -1,2 +1 @@
...
-audit = false
+save-exact = true
```
Comment (bristol-mike-jones) [inline .npmrc:1]: how come we've removed ```audit-level = high``` ?

File: src/models/adr-details.ts [hunk 20]
```diff
@@ -0,0 +1,179 @@
+import {DynamoDbImage, parseStringArray} from "../services/dynamodb-images";
+import {Maybe} from "./optionals";
+
+export interface AdrDetails {
+    vehicleDetails?: VehicleDetails;
+    listStatementApplicable?: boolean;
+    batteryListNumber?: string;
+    declarationsSeen?: boolean;
+    brakeDeclarationsSeen?: boolean;
```
[the remark this answers is not in the document]
Comment (PaulBenn-BJSS) [reply src/models/adr-details.ts:9]: As agreed on chat with Jon, leaving this for a future PR and resolving for now - @SanjeetBJSS accepts the tech debt

File: src/models/shared-enums.ts (hunk 12 above)
Comment (PaulBenn-BJSS) [reply src/models/shared-enums.ts:3]: As agreed on chat with Jon, leaving this for a future PR and resolving for now - @SanjeetBJSS accepts the tech debt

File: src/models/tech-record-document.ts (hunk 13 above)
Comment (PaulBenn-BJSS) [reply src/models/tech-record-document.ts:5]: As agreed on chat with Jon, leaving this for a future PR and resolving for now - @SanjeetBJSS accepts the tech debt

File: src/models/adr-details.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (bristol-mike-jones) [reply src/models/adr-details.ts:9]: Yep, having a common interface that we share across repo's would be great.  I think it's a bit of a stretch to make this an ask in a PR. There are a whole host of things we need to think about first.  Probably a ticket for the backlog?

File: package.json (hunk 6 above)
Comment (PaulBenn-BJSS) [reply package.json:31]: resolving this - Jon has double-checked current config runs the proper auditing process in the pipeline

File: .npmrc (hunk 19 above)
Comment (PaulBenn-BJSS) [reply .npmrc:1]: resolving this - Jon has now double-checked current config runs the proper auditing process in the pipeline :)

File: .gitmodules (hunk 2 above)
Comment (sdh100shaun) [reply .gitmodules:1]: Have others in development worked on this and are comfortable it is workable - if @JonLazariniDD  etc say that is fine. I am OK if it is working

Comment (JonLazariniDD) [reply .gitmodules:1]: Paul & I had a catch up yesterday about this and clarified in the README an additional step (where the branch of the sql repo with the changes had to be checked out and I had this step wrong).
I think it has been implemented (and automated) in the pipeline so it can be a good approach for now and use an alternative approach later should we think it is required.

File: package.json (hunk 7 above)
Comment (JonLazariniDD) [reply package.json:28]: thanks for clarifying, I appreciate and feel free to resolve guys if you think it is ok.

**Review by SanjeetBJSS — APPROVED** [review]
Happy to approve.
